File: kernel/include/kernel/io.c
#include <kernel/io.h>

#include <stdint.h>

inline void insl(int port, void *addr, int cnt) {
    asm volatile("cld; rep insl" :
                "=D" (addr), "=c" (cnt) :
                "d" (port), "0" (addr), "1" (cnt) :
                "memory", "cc"
    );
}

inline uint8_t inb(uint16_t port) {
    uint8_t return_value;
    asm volatile ( "inb %1, %0" : "=a"(return_value) : "Nd"(port) );
    return return_value;
}

inline uint8_t inw(uint16_t port) {
    uint16_t return_value;
    asm volatile ( "inw %1, %0" : "=a"(return_value) : "Nd"(port) );
    return return_value;
}

inline uint8_t inl(uint16_t port) {
    uint32_t return_value;
    asm volatile ( "inl %1, %0" : "=a"(return_value) : "Nd"(port) );
    return return_value;
}

inline void outsl(int port, const void *addr, int cnt) {
    asm volatile("cld; rep outsl" :
                "=S" (addr), "=c" (cnt) :
                "d" (port), "0" (addr), "1" (cnt) :
                "cc"
    );
}

inline void outb(uint16_t port, uint8_t val) {
    asm volatile ( "outb %0, %1" : : "a"(val), "Nd"(port) );
}

inline void outw(uint16_t port, uint16_t val) {
    asm volatile ( "outw %0, %1" : : "a"(val), "Nd"(port) );
}

inline void outl(uint16_t port, uint32_t val) {
    asm volatile ( "outl %0, %1" : : "a"(val), "Nd"(port) );
}

File: kernel/src/io.zig
pub inline fn inb(port: u16) u8 {
    return asm volatile ("inb %[port], %[result]"
        : [result] "={al}" (-> u8),
        : [port] "N{dx}" (port),
    );
}

pub inline fn insl(port: u16, address: *void, count: u32) void {
    asm volatile ("cld; rep insl"
        : [address1] "={edi}" (address),
          [count1] "={ecx}" (count),
        : [port] "{edx}" (port),
          [address2] "0" (address),
          [count2] "1" (count),
        : "memory", "cc"
    );
}

pub inline fn outb(port: u16, value: u8) void {
    asm volatile ("outb %[value], %[port]"
        :
        : [value] "{al}" (value),
          [port] "N{dx}" (port),
    );
}

pub inline fn outsl(port: u16, address: *void, count: u32) void {
    asm volatile ("cld; rep outsl"
        : [address1] "={esi}" (address),
          [count1] "={ecx}" (count),
        : [port] "{edx}" (port),
          [address2] "0" (address),
          [count2] "1" (count),
        : "cc"
    );
}
